### PR TITLE
Dynamic Group performance improvements for relationships in filtersets.

### PR DESCRIPTION
**Pull request by jathanism**
# Related to: #3269
# What's Changed

**Some of my assertions may be incorrect. Please keep me honest.**

This change is related to improving the way Dynamic Groups members are collected since they are built around filtersets. In doing some profiling it was determined that the table for relationships is hit twice, with one (1) query to the table for content types as well, resulting in three (3) queries each time a Dynamic Group is constructed from a filterset.

With these changes, this reduces the construction of a filterset that relationships down to a single query.

- Changed `RelationshipModelFilterSetMixin` to perform a single OR query including `select_related` for `source_type` and `destination_type` vs. two single queries for each source/destination types.

### Before

This query is performed once for each source/destination plus one for content type:

```
14:39:03.078 DEBUG   django.db.backends   utils.py                             debug_sql() :
  (0.008) SELECT "extras_relationship"."id", "extras_relationship"."created", "extras_relationship"."last_updated", "extras_relationship"."name", "extras_relationship"."slug", "extras_relationship"."description", "extras_relationship"."type", "extras_relationship"."required_on", "extras_relationship"."source_type_id", "extras_relationship"."source_label", "extras_relationship"."source_hidden", "extras_relationship"."source_filter", "extras_relationship"."destination_type_id", "extras_relationship"."destination_label", "extras_relationship"."destination_hidden", "extras_relationship"."destination_filter", "extras_relationship"."advanced_ui" FROM "extras_relationship" WHERE (NOT "extras_relationship"."source_hidden" AND "extras_relationship"."source_type_id" = 3) ORDER BY "extras_relationship"."name" ASC; args=(3,)
14:39:03.079 DEBUG   django.db.backends   utils.py                             debug_sql() :
  (0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 11 LIMIT 21; args=(11,)
14:39:03.080 DEBUG   django.db.backends   utils.py                             debug_sql() :
  (0.000) SELECT "extras_relationship"."id", "extras_relationship"."created", "extras_relationship"."last_updated", "extras_relationship"."name", "extras_relationship"."slug", "extras_relationship"."description", "extras_relationship"."type", "extras_relationship"."required_on", "extras_relationship"."source_type_id", "extras_relationship"."source_label", "extras_relationship"."source_hidden", "extras_relationship"."source_filter", "extras_relationship"."destination_type_id", "extras_relationship"."destination_label", "extras_relationship"."destination_hidden", "extras_relationship"."destination_filter", "extras_relationship"."advanced_ui" FROM "extras_relationship" WHERE (NOT "extras_relationship"."destination_hidden" AND "extras_relationship"."destination_type_id" = 3) ORDER BY "extras_relationship"."name" ASC; args=(3,)
```

Analysis:

- Source: https://explain.depesz.com/s/Enul
- Destination: https://explain.depesz.com/s/Qqkj
- Content type: https://explain.depesz.com/s/idys

### After

This query is performed combining source/destination:

```
14:41:24.598 DEBUG   django.db.backends   utils.py                             debug_sql() :
  (0.014) DECLARE "_django_curs_8739029312_sync_1" NO SCROLL CURSOR WITH HOLD FOR SELECT "extras_relationship"."id", "extras_relationship"."created", "extras_relationship"."last_updated", "extras_relationship"."name", "extras_relationship"."slug", "extras_relationship"."description", "extras_relationship"."type", "extras_relationship"."required_on", "extras_relationship"."source_type_id", "extras_relationship"."source_label", "extras_relationship"."source_hidden", "extras_relationship"."source_filter", "extras_relationship"."destination_type_id", "extras_relationship"."destination_label", "extras_relationship"."destination_hidden", "extras_relationship"."destination_filter", "extras_relationship"."advanced_ui", "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model", T3."id", T3."app_label", T3."model" FROM "extras_relationship" INNER JOIN "django_content_type" ON ("extras_relationship"."source_type_id" = "django_content_type"."id") INNER JOIN "django_content_type" T3 ON ("extras_relationship"."destination_type_id" = T3."id") WHERE ((NOT "extras_relationship"."source_hidden" AND "extras_relationship"."source_type_id" = 3) OR (NOT "extras_relationship"."destination_hidden" AND "extras_relationship"."destination_type_id" = 3)) ORDER BY "extras_relationship"."name" ASC; args=(3, 3)
```

Analysis:

https://explain.depesz.com/s/xR8z

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/filters/mixins.py [hunk 1]
```diff
@@ -219,11 +219,16 @@ def _append_relationships(self, model):
...
-        self._append_relationships_side(dest_relationships, RelationshipSideChoices.SIDE_DESTINATION, model)
+        query = Q(source_type=self.obj_type, source_hidden=False) | Q(
+            destination_type=self.obj_type, destination_hidden=False
+        )
+        relationships = Relationship.objects.select_related("source_type", "destination_type").filter(query)
+
+        for rel in relationships.iterator():
+            if rel.source_type == self.obj_type:
+                self._append_relationships_side([rel], RelationshipSideChoices.SIDE_SOURCE, model)
+            if rel.destination_type == self.obj_type:
```
Comment (glennmatthews) [inline nautobot/extras/filters/mixins.py:230]: Maybe?

```suggestion
            if rel.source_type == self.obj_type and not rel.source_hidden:
                self._append_relationships_side([rel], RelationshipSideChoices.SIDE_SOURCE, model)
            if rel.destination_type == self.obj_type and not rel.destination_hidden:
```

File: nautobot/extras/migrations/0055_extras__relationships__index_together.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Generated by Django 3.2.18 on 2023-02-24 21:38
```
Comment (glennmatthews) [inline nautobot/extras/migrations/0055_extras__relationships__index_together.py:1]: Mmm, more migrations to renumber next time we merge up into `next`... :-)

File: nautobot/extras/models/relationships.py [hunk 3]
```diff
@@ -709,6 +713,10 @@ class Meta:
             "destination_type",
             "destination_id",
         )
+        index_together = [
```
Comment (glennmatthews) [inline nautobot/extras/models/relationships.py:716]: Mind adding a comment explaining the usefulness of these two indexes? They're not as self-explanatory as the ones above.

File: nautobot/extras/migrations/0055_extras__relationships__index_together.py (hunk 2 above)
Comment (jathanism) [reply nautobot/extras/migrations/0055_extras__relationships__index_together.py:1]: 😑

File: nautobot/extras/filters/mixins.py (hunk 1 above)
Comment (jathanism) [reply nautobot/extras/filters/mixins.py:230]: That is already covered in the query, no? Anything with `{side}_hidden` should not be in the queryset. See:

```python
relationships = Relationship.objects.select_related("source_type", "destination_type").filter(query)
```

File: nautobot/extras/models/relationships.py (hunk 3 above)
Comment (jathanism) [reply nautobot/extras/models/relationships.py:716]: Yeah right now it's actually looking like the indexes haven't made much of a difference and I may be wrong here.

**Review by bryanculver — COMMENTED** [review]
How much do the query optimizations (single filter vs two filters) get us vs the index_together? Just curious to avoid a reordered migration in `next`.

**Comment by jathanism** [comment]
> How much do the query optimizations (single filter vs two filters) get us vs the index_together? Just curious to avoid a reordered migration in `next`.

On secondary review, I don't think the indexes have made any changes at all here. At least not on deeper inspection of the explain queries analysis.

File: nautobot/extras/filters/mixins.py (hunk 1 above)
Comment (jathanism) [reply nautobot/extras/filters/mixins.py:230]: Gotcha. Thanks for clarifying.

**Comment by jathanism** [comment]
> > How much do the query optimizations (single filter vs two filters) get us vs the index_together? Just curious to avoid a reordered migration in `next`.
> 
> On secondary review, I don't think the indexes have made any changes at all here. At least not on deeper inspection of the explain queries analysis.

I ripped out the new indexes, as I feel my initial assertions about them working was incorrect.

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**